### PR TITLE
Use accept4(2) on platforms that support it

### LINKED ISSUE #1213: Use accept4 where possible
In `sys::TcpListener::accept`.

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**
Fixes #1213.

~~Depends on #1242.~~

### CODE REVIEW

File: src/sys/windows/tcp.rs [hunk 1]
```diff
@@ -64,3 +64,8 @@ pub fn bind(addr: SocketAddr) -> io::Result<net::TcpListener> {
         .map(|_| unsafe { net::TcpListener::from_raw_socket(socket as StdSocket) })
     })
 }
+
+pub fn accept(listener: &net::TcpListener) -> io::Result<(net::TcpStream, SocketAddr)> {
+    // The non-blocking state of `listener` is inherited.
```
Comment (kleimkuhler) [inline src/sys/windows/tcp.rs]: Is the source for this assumption the [Remark](https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-accept#remarks)? I don't think the source needs to be an explicit comment, but just want to make sure this comment isn't assuming from something else.

File: src/sys/unix/tcp.rs [hunk 2]
```diff
@@ -50,3 +50,52 @@ pub fn bind(addr: SocketAddr) -> io::Result<net::TcpListener> {
...
+            addr.as_mut_ptr() as *mut _,
+            &mut length,
+            libc::SOCK_CLOEXEC | libc::SOCK_NONBLOCK,
+        ))
+        .map(|socket| unsafe { net::TcpStream::from_raw_fd(socket) })
+    }?;
+
+    // But not all platforms have the `accept4(2)` call. Luckily BSD (derived)
+    // OSes inherit the non-blocking flag from the listener, so we just have to
+    // set `CLOEXEC`.
```
Comment (kleimkuhler) [inline src/sys/unix/tcp.rs]: > When using  accept(), portable programs should not rely on the O_NONBLOCK
> and O_ASYNC properties and  the signal destination being inherited, but
> should set them explicitly using fcntl(2); accept4() sets these properties
> consistently, but may not be fully portable across UNIX platforms.

This makes sense. The "portable" programs are eiter going to have `accept4`, or we can assume inherit `NONBLOCK`.

File: src/sys/windows/tcp.rs (hunk 1 above)
Comment (Thomasdezeeuw) [reply src/sys/windows/tcp.rs]: I was indeed going from those docs, I've added a link in the last commit anyway. Always useful for people (and future me when I forget).

File: src/sys/unix/tcp.rs (hunk 2 above)
Comment (Thomasdezeeuw) [reply src/sys/unix/tcp.rs]: I've manually checked all the OSes and the list and they all inherit the `NONBLOCK` property, it seems only Linux really diverges from this.

**Pull request merged**